packages/core/lib/attributes.ts

```diff
@@ -34,15 +35,64 @@ export interface SpanAttributesSource <C extends Configuration> {
   requestAttributes: (span: SpanInternal) => void
 }
 
+export interface SpanAttributesLimits {
+  attributeStringValueLimit: number
+  attributeArrayLengthLimit: number
+  attributeCountLimit: number
+}
+
+function truncateString (value: string, limit: number) {
+  const originalLength = value.length
+  const newString = value.slice(0, limit)
+  const truncatedLength = newString.length
+
+  return `${newString} *** ${originalLength - truncatedLength} CHARS TRUNCATED`
+}
+
 export class SpanAttributes {
   private readonly attributes: Map<string, SpanAttribute>
+  private readonly logger: Logger
+  private readonly spanAttributeLimits: SpanAttributesLimits
+  private _droppedAttributesCount = 0
+
+  get droppedAttributesCount () {
+    return this._droppedAttributesCount
+  }
 
-  constructor (initialValues: Map<string, SpanAttribute>) {
+  constructor (initialValues: Map<string, SpanAttribute>, spanAttributeLimits: SpanAttributesLimits, logger: Logger) {
     this.attributes = initialValues
+    this.spanAttributeLimits = spanAttributeLimits
+    this.logger = logger
+  }
+
+  // Validate after the span has ended and after the sampling
+  private validateAttribute (name: string, value: SpanAttribute) {
+    if (name.length > ATTRIBUTE_KEY_LENGTH_LIMIT) {
+      this.logger.warn('Attribute key limit reached. Discarding attribute.')
```

Comment:
I would be keen to update this message wording to something like Dropping attribute with key "really-long-key". Key length exceeds limit of 128 characters.